### ISSUE TITLE
"social share" (???) images for separate pages?

### ISSUE THREAD

**Issue by nycnoise** [gregsadetsky/nycnoise#249]
wondering if there's a way to set the images that show up when sharing different pages –– i think this is called "social share" ?? –– on wordpress i just made tiny images that said, e.g., SITES LIKE THIS or whatever & had them at the top of the page so it would show up when sending page as a text, but i imagine there's a smarter way to do this...

**Comment by gregsadetsky** [comment]
Yeaahhh I was exactly righht just now looking at this, because in general I want to revamp our opengraph / social share thing

This concerns #206 and #247

But yeah briefly speaking, yes. We could have an image upload for static pages (in the admin) and those would show up in the opengraph i.e. when a page is shared

**Comment by nycnoise** [comment]
this would rule (i've been sharing resources w/ people in chats & it's getting a little hairy lol) –– but also would even be fine if preview had the same image, but showed the **page name as text**? (if that's any easier...?)

**Comment by gregsadetsky** [comment]
For sure that’s version 0 and would be the default for all events. Even if we add images, not all shows will have images. So yes. The share title would be the event textOn Mar 9, 2024, at 3:25 PM, nyc noise ***@***.***> wrote:﻿
this would rule (i've been sharing resources w/ people in chats & it's getting a little hairy lol) –– but also would even be fine if preview had the same image, but showed the page name as text? (if that's any easier...?)

**Comment by litviakk** [comment]
this might be tricky - I assume we don't want these images to be served from db, and we might want to utilize compression via whitenoise, therefore we might have to dump them to `static` dir and either reload the server or run collectstatic on updates

**Comment by gregsadetsky** [comment]
for static pages, we could have a file upload file in the models and store the files in a S3 bucket

using https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html it's pretty easy to setup

django-storages gives you a file upload field in the django admin, files are then uploaded to s3, and then they can be referenced in the templates from the model (i.e. `{{ static_page.image }}`) and then the correct s3 url is inserted

**Comment by litviakk** [comment]
we could, but honestly, I'd keep all static files in the same place (maybe even include static HTML files there) for simplicity's sake
seemingly it's just more hustle to set up IaC for s3 + their CDN (which might be pricey if we have some heavy traffic spikes) than to run `call_command('collectstatic')` on file uploads from admin + it should be quite fast since it will only update new files
but I might be missing something

**Comment by gregsadetsky** [comment]
I wouldn't worry about it - django-storages and s3 is easy to setup, and no need for a cdn (like cloudflare or cloudfront), even though those are truly not that bad at all. I've done it with many django projects. (also s3/aws has a generous free tier)

the bigger point for this specific issue is that Jessica (who runs the site) wants to "set the images that show up when sharing different pages" i.e. she wants a dynamic way (via the django admin) to set/change the images that are relevant for each specific static page. so it's not something that can be done via collectstatic unfortunately

let me know if this makes sense

**Comment by litviakk** [comment]
yes, that's exactly how I see it - uploading images through admin
let me try drafting it up and if I run into something we can run w/ s3, how does it sound to you?

**Comment by gregsadetsky** [comment]
for sure, sounds great! let me know if I can help with the s3 setup

roughly speaking you setup an aws account, then setup an account in IAM (which can only access S3 - it's a tiny bit better that way), create an AWS access + secret key credentials, configure django-storages, and that's pretty much it. you also need to make the S3 bucket public so that the URLs to the files are publicly accessible.

and I'd make a dev and a prod s3 bucket so that it's easier to test/develop. (you'll need test + prod aws credentials for this i.e. two IAM accounts)

**Comment by litviakk** [comment]
Just pushed a [draft](https://github.com/gregsadetsky/nycnoise/pull/321) w/ collectstatic, it's working seemingly fine.
The last thing to add it to opengraph tags, but it's easier to do after we migrate to django-meta

**Comment by gregsadetsky** [comment]
I mean yeah this wouldn't work on production in a cloud (like heroku/render) environment -- files on the server are re-created every time we git push, and collectstatic is run once when deploying code. we wouldn't want to (or couldn't even) run collectstatic once the server is deployed. files are temporary on a deployed instance.

this is why we need to store files elsewhere (like s3)

**Comment by litviakk** [comment]
make sense, we could keep the images in the db and dump them on the disk on deploy

anyway, we can go w/ s3, I don't see a problem there, other than keeping static files all around the internet
can you set up an AWS account? I would rather create a [role](https://aws.amazon.com/blogs/security/use-iam-roles-to-connect-github-actions-to-actions-in-aws/) for github actions so that it can create buckets and roles via IaC, and propagate everything required to the prod env
for local development there is also [localstack](https://github.com/localstack/localstack)

**Comment by gregsadetsky** [comment]
yeah for sure I'll setup aws and provide you with the access key offline

I need to get to a few projects/things that I'm behind on. and maybe let's start with the rest of your PRs too first just so that there's not too many things in flight

but yeah having django-storages setup and s3 will be good and most probably useful again in the future (there was talk at some point about hosting show posters - and those are images too..!)

**Comment by gregsadetsky** [comment]
note that we are discussing this here: https://github.com/gregsadetsky/nycnoise/pull/321
